File: sdk/python/examples/langgraph/31_classify_and_route.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Classify and Route — create_agent with domain-specialist tools.

Demonstrates:
    - Using create_agent with a system prompt that directs classification and routing
    - Domain-specialist tools (science, history, sports, technology, cooking)
    - The LLM classifies the input and routes to the right specialist tool
    - Practical use case: smart help desk that routes to the right department

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def answer_science(question: str) -> str:
    """Answer a science question with precise, relevant scientific context.

    Args:
        question: The science question to answer.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a science expert. Answer precisely with relevant scientific context."),
        ("human", "{question}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"question": question})
    return f"[Science Expert] {response.content.strip()}"


@tool
def answer_history(question: str) -> str:
    """Answer a history question with historical context and key dates.

    Args:
        question: The history question to answer.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a history expert. Provide historical context and key dates."),
        ("human", "{question}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"question": question})
    return f"[History Expert] {response.content.strip()}"


@tool
def answer_sports(question: str) -> str:
    """Answer a sports question with stats and context when relevant.

    Args:
        question: The sports question to answer.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a sports analyst. Give stats and context when relevant."),
        ("human", "{question}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"question": question})
    return f"[Sports Analyst] {response.content.strip()}"


@tool
def answer_technology(question: str) -> str:
    """Answer a technology question clearly and with technical accuracy.

    Args:
        question: The technology question to answer.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a technology expert. Be clear and technically accurate."),
        ("human", "{question}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"question": question})
    return f"[Tech Expert] {response.content.strip()}"


@tool
def answer_cooking(question: str) -> str:
    """Answer a cooking question with practical, delicious advice.

    Args:
        question: The cooking question to answer.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a professional chef. Give practical, delicious advice."),
        ("human", "{question}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"question": question})
    return f"[Chef] {response.content.strip()}"


ROUTER_SYSTEM = """You are a smart question router.

Classify each question into one of these categories, then call the matching tool:
- Science questions (physics, chemistry, biology, etc.) → answer_science
- History questions (events, dates, historical figures) → answer_history
- Sports questions (games, athletes, stats) → answer_sports
- Technology questions (software, hardware, computing) → answer_technology
- Cooking questions (recipes, ingredients, techniques) → answer_cooking

Call exactly one specialist tool and return its response.
"""

graph = create_agent(
    llm,
    tools=[answer_science, answer_history, answer_sports, answer_technology, answer_cooking],
    name="classify_and_route_agent",
    system_prompt=ROUTER_SYSTEM,
)

if __name__ == "__main__":
    questions = [
        "What is photosynthesis?",
        "When did World War II end?",
        "Who has won the most Grand Slam tennis titles?",
        "What is Kubernetes?",
        "How do I make a perfect risotto?",
    ]

    with AgentRuntime() as runtime:
        for q in questions:
            print(f"\nQ: {q}")
            result = runtime.run(graph, q)
            result.print_result()

File: sdk/typescript/examples/langgraph/31-classify-and-route.ts
/**
 * Classify and Route -- LLM-based input classification with specialized routing.
 *
 * Demonstrates:
 *   - Using an LLM to classify input into a discrete category
 *   - Conditional edges routing to specialized handler nodes
 *   - Each handler node is tailored to its domain
 *   - Practical use case: smart help desk that routes to the right department
 */

import { StateGraph, START, END, Annotation } from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// LLM
// ---------------------------------------------------------------------------
const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });

// ---------------------------------------------------------------------------
// State schema
// ---------------------------------------------------------------------------
const ClassifyState = Annotation.Root({
  input: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  category: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  answer: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type State = typeof ClassifyState.State;

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
async function classify(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      'Classify the input into exactly one category. ' +
        'Categories: science, history, sports, technology, cooking. ' +
        'Respond with the category name only.',
    ),
    new HumanMessage(state.input),
  ]);
  return { category: String(response.content).trim().toLowerCase() };
}

async function answerScience(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('You are a science expert. Answer precisely with relevant scientific context.'),
    new HumanMessage(state.input),
  ]);
  return { answer: `[Science Expert] ${String(response.content).trim()}` };
}

async function answerHistory(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('You are a history expert. Provide historical context and key dates.'),
    new HumanMessage(state.input),
  ]);
  return { answer: `[History Expert] ${String(response.content).trim()}` };
}

async function answerSports(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('You are a sports analyst. Give stats and context when relevant.'),
    new HumanMessage(state.input),
  ]);
  return { answer: `[Sports Analyst] ${String(response.content).trim()}` };
}

async function answerTechnology(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('You are a technology expert. Be clear and technically accurate.'),
    new HumanMessage(state.input),
  ]);
  return { answer: `[Tech Expert] ${String(response.content).trim()}` };
}

async function answerCooking(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage('You are a professional chef. Give practical, delicious advice.'),
    new HumanMessage(state.input),
  ]);
  return { answer: `[Chef] ${String(response.content).trim()}` };
}

function route(state: State): string {
  const mapping: Record<string, string> = {
    science: 'science',
    history: 'history',
    sports: 'sports',
    technology: 'technology',
    cooking: 'cooking',
  };
  return mapping[state.category || ''] || 'technology'; // default to technology
}

// ---------------------------------------------------------------------------
// Build the graph
// ---------------------------------------------------------------------------
const builder = new StateGraph(ClassifyState);
builder.addNode('classify', classify);
builder.addNode('science', answerScience);
builder.addNode('history', answerHistory);
builder.addNode('sports', answerSports);
builder.addNode('technology', answerTechnology);
builder.addNode('cooking', answerCooking);

builder.addEdge(START, 'classify');
builder.addConditionalEdges('classify', route, {
  science: 'science',
  history: 'history',
  sports: 'sports',
  technology: 'technology',
  cooking: 'cooking',
});
builder.addEdge('science', END);
builder.addEdge('history', END);
builder.addEdge('sports', END);
builder.addEdge('technology', END);
builder.addEdge('cooking', END);

const graph = builder.compile();

// Add agentspan metadata for extraction
(graph as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  tools: [],
  framework: 'langgraph',
};

const PROMPT = 'What is photosynthesis?';

// ---------------------------------------------------------------------------
// Run on agentspan
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(graph, PROMPT);
    console.log('Status:', result.status);
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(graph);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples/langgraph --agents classify_and_route
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(graph);
  } finally {
    await runtime.shutdown();
  }
}

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('31-classify-and-route.ts') || process.argv[1]?.endsWith('31-classify-and-route.js')) {
  main().catch(console.error);
}
